distributed/tests/test_active_memory_manager.py

```diff
@@ -903,6 +906,88 @@ async def test_RetireWorker_all_recipients_are_paused(c, s, a, b):
     assert await c.submit(inc, 1) == 2
 
 
+@gen_cluster(
+    client=True,
+    config={
+        "distributed.scheduler.active-memory-manager.start": True,
+        "distributed.scheduler.active-memory-manager.policies": [],
+    },
+    timeout=15,
+)
+async def test_RetireWorker_new_keys_arrive_after_all_keys_moved_away(
+    c, s: Scheduler, a: Worker, b: Worker
+):
+    """
+    If all keys have been moved off a worker, but then new keys arrive (due to task completion or `gather_dep`)
+    before the worker has actually closed, make sure we still retire it (instead of hanging forever).
+
+    This test is timing-sensitive. If it runs too slowly, it *should* `pytest.skip` itself.
+    However, this won't work until https://github.com/dask/distributed/issues/6239 is fixed.
+    So it may get stuck and fail at the 15s test timeout.
+
+    See https://github.com/dask/distributed/issues/6223 for motivation.
+    """
+    ws_a = s.workers[a.address]
+    ws_b = s.workers[b.address]
+    event = Event()
+
+    # Put 200 keys on the worker, so `_track_retire_worker` will sleep for 0.5s
+    # (Note that if you decrease the 200 to a very small value (say 2), this test will time out.
+    # But just because `extra` is restricted to only run on worker A, and A gets closed before `extra` can run,
+    # so not a deadlock (just user error)).
+    xs = c.map(lambda x: x, range(200), workers=[a.address])
+
+    while len(a.data) < len(xs):
+        await asyncio.sleep(0.1)
+
+    extra = c.submit(lambda: print(event.wait("2s")), workers=[a.address])
+
+    while extra.key not in a.tasks or a.tasks[extra.key].state != "executing":
+        await asyncio.sleep(0.01)
+
+    t = asyncio.create_task(c.retire_workers([a.address]))
+
+    # Wait for all `xs` to be replicated.
+    while not len(ws_b.has_what) == len(xs):
+        await asyncio.sleep(0.01)
+
+    # `_track_retire_worker` _should_ now be sleeping for 0.5s, because there were >=200 keys on A.
+    # In this test, everything from here on needs to happen within 0.5s.
+
+    # The policy runs again. Because the default 2s AMM interval is 4x longer than the 0.5s wait,
+    # what we're about to trigger is unlikely, but still theoretically possible for the times to line up.
+    amm: ActiveMemoryManagerExtension = s.extensions["amm"]
+    assert len(amm.policies) == 1
+    policy = next(iter(amm.policies))
+    assert isinstance(policy, RetireWorker)
+
+    # This will drop all the `xs` from A (since they're already replicated on B).
+    amm.run_once()
```

Comment:
Unnecessary. Dropping is not a precondition for done; it's just to reduce memory pressure in case of spilled keys.

Reply:
Calling `run_once` is necessary, because what we wan to test is:

* `_track_retire_worker` is sleeping
* policy runs and removes itself because all keys have been replicated
* another key appears on the retiring worker
* `_track_retire_worker` wakes up

By running it once here, it will remove itself.